homeassistant/components/numato/__init__.py

```diff
@@ -0,0 +1,248 @@
+    if not (isinstance(rng[0], int) and isinstance(rng[1], int)):
+        raise vol.Invalid(f"Only integers are allowed: {rng}")
+    if len(rng) != 2:
+        raise vol.Invalid(f"Only two numbers allowed in a range: {rng}")
+    if rng[0] > rng[1]:
+        raise vol.Invalid(f"Lower range bound must come first: {rng}")
+    return rng
+
+
+def float_range(rng):
+    """Validate the input array to describe a range by two floats."""
+    try:
+        coe = vol.Coerce(float)
+        coe(rng[0])
+        coe(rng[1])
+    except vol.CoerceInvalid:
+        raise vol.Invalid(f"Only int or float values are allowed: {rng}")
+    if len(rng) != 2:
+        raise vol.Invalid(f"Only two numbers allowed in a range: {rng}")
+    if rng[0] > rng[1]:
+        raise vol.Invalid(f"Lower range bound must come first: {rng}")
+    return rng
+
+
+def adc_port_number(num):
+    """Validate input number to be in the range of ADC enabled ports."""
+    try:
+        num = int(num)
+    except (ValueError):
+        raise vol.Invalid(f"Port numbers must be integers: {num}")
+    if num not in range(1, 8):
+        raise vol.Invalid(f"Only port numbers from 1 to 7 are ADC capable: {num}")
+    return num
+
+
+ADC_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Optional(CONF_SRC_RANGE, default=DEFAULT_SRC_RANGE): int_range,
+        vol.Optional(CONF_DST_RANGE, default=DEFAULT_DST_RANGE): float_range,
+        vol.Optional(CONF_DST_UNIT, default=DEFAULT_UNIT): cv.string,
+    }
+)
+
+PORTS_SCHEMA = vol.Schema({cv.positive_int: cv.string})
+
+IO_PORTS_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_PORTS): PORTS_SCHEMA,
+        vol.Optional(CONF_INVERT_LOGIC, default=DEFAULT_INVERT_LOGIC): cv.boolean,
+    }
+)
+
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ID): cv.positive_int,
+        CONF_BINARY_SENSORS: IO_PORTS_SCHEMA,
+        CONF_SWITCHES: IO_PORTS_SCHEMA,
+        CONF_SENSORS: {CONF_PORTS: {adc_port_number: ADC_SCHEMA}},
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: {
+            CONF_DEVICES: vol.All(cv.ensure_list, [DEVICE_SCHEMA]),
+            vol.Optional(CONF_DISCOVER, default=DEFAULT_DEV): vol.All(
+                cv.ensure_list, [cv.string]
+            ),
+        },
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+def setup(hass, config):
+    """Initialize the numato integration.
+
+    Discovers available Numato devices and loads the binary_sensor, sensor and
+    switch platforms.
+
+    Returns False on error during device discovery (e.g. duplicate ID),
+    otherwise returns True.
+
+    No exceptions should occur, since the platforms are initialized on a best
+    effort basis, which means, errors are handled locally.
+    """
+    hass.data[DOMAIN] = config[DOMAIN]
+
+    try:
+        gpio.discover(config[DOMAIN][CONF_DISCOVER])
+    except gpio.NumatoGpioError as err:
+        _LOGGER.info("Error discovering Numato devices: %s", str(err))
```

Comment:
We don't need to copy the exception argument to string in log messages. The exception argument has a magic method `__str__`.
```suggestion
        _LOGGER.info("Error discovering Numato devices: %s", err)
```

Reply:
Sure, my bad. Fixed all in a single commit, when I saw, that you had prepared suggestions to take over. Sorry!

Reply:
No worries. From the files tab in github PR view you can batch all suggestions into one commit before committing the suggestions.